python/ray/data/_internal/logical/interfaces/operator.py

```diff
@@ -10,7 +16,7 @@ class Operator:
     def __init__(
         self,
         name: str,
-        input_dependencies: List["Operator"],
+        input_dependencies: List[Self],
```

Comment:
Not super familiar with the `Self` type annotation so could be totally mistaken, but I think this might be slightly incorrect? For example, `input_dependencies` of a physical `MapOperator` should be a `list[PhysicalOperator]`, but a type checker would expect a `list[MapOperator]`?

Reply:
It only depends on the type hints for the variable, regardless of the actual type of the object.
E.g. in most of our code, we annotate the variables with `PhysicalOperator`, instead of `MapOperator`. 

```python
def foo(op: PhysicalOpeartor):
    for next_op in op.output_dependencies:
          ...  # mypy will think next_op is a PhysicalOperator here. 
```

Reply:
it's not ideal though. maybe I can try overriding `op.output_dependencies` in `PhysicalOperator`

Reply:
updated to use overriding. this also eliminates the type hinting errors.